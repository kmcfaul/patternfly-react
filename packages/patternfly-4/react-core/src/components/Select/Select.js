import React from 'react';
import styles from '@patternfly/patternfly/components/Select/select.css';
import { css } from '@patternfly/react-styles';
import PropTypes from 'prop-types';
import SingleSelect from './SingleSelect';
import SelectToggle from './SelectToggle';
import { SelectContext } from './selectConstants';

// seed for the aria-labelledby ID
let currentId = 0;

const propTypes = {
  /** Content rendered inside the Select */
  children: PropTypes.node,
  /** Classes applied to the root of the Select */
  className: PropTypes.string,
  /** Flag to indicate if select is expanded */
  isExpanded: PropTypes.bool,
  /** Placeholder text of Select */
  placeholderText: PropTypes.string,
  /** Array of SelectOption nodes that will be rendered */
  selectOptions: PropTypes.array,
  /** Selected item(s) structure */
  selections: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  /** Callback for selection behavior */
  onSelect: PropTypes.func.isRequired,
  /** Callback for toggle button behavior */
  onToggle: PropTypes.func.isRequired,
  /** Variant of rendered Select */
  variant: PropTypes.oneOf(['single']),
  /** Width of the select container as a number of px or string percentage */
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Additional props are spread to the container <ul> */
  '': PropTypes.any
};

const defaultProps = {
  children: null,
  className: '',
  isExpanded: false,
  selectOptions: null,
  selections: null,
  placeholderText: null,
  variant: 'single',
  width: '100%'
};

class Select extends React.Component {
  parentRef = React.createRef();
  state = { openedOnEnter: false };

  onEnter = () => {
    this.setState({ openedOnEnter: true });
  };

  onClose = () => {
    this.setState({ openedOnEnter: false });
  };

  render() {
    const {
      children,
      className,
      variant,
      onToggle,
      onSelect,
      isExpanded,
      selectOptions,
      selections,
      placeholderText,
      width,
      ...props
    } = this.props;
    const { openedOnEnter } = this.state;
    const renderedChildren = children || selectOptions;
    let childPlaceholderText = null;
    if (!selections && !placeholderText) {
      const childPlaceholder = renderedChildren.filter(child => child.props.isPlaceholder === true);
      childPlaceholderText =
        (childPlaceholder[0] && childPlaceholder[0].props.value) ||
        (renderedChildren[0] && renderedChildren[0].props.value);
    }

    return (
      <div
        className={css(styles.select, isExpanded && styles.modifiers.expanded, className)}
        ref={this.parentRef}
        style={{ width }}
      >
        <SelectContext.Provider value={onSelect}>
          {variant === 'single' && (
            <React.Fragment>
              <SelectToggle
                id={`pf-toggle-id-${currentId++}`}
                parentRef={this.parentRef.current}
                isExpanded={isExpanded}
                onToggle={onToggle}
                onEnter={this.onEnter}
                onClose={this.onClose}
              >
                {selections || placeholderText || childPlaceholderText}
              </SelectToggle>
              {isExpanded && (
                <SingleSelect {...props} selected={selections} openedOnEnter={openedOnEnter}>
                  {renderedChildren}
                </SingleSelect>
              )}
            </React.Fragment>
          )}
        </SelectContext.Provider>
      </div>
    );
  }
}

Select.propTypes = propTypes;
Select.defaultProps = defaultProps;

export default Select;
