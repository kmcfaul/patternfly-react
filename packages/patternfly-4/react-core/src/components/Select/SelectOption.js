import React from 'react';
import styles from '@patternfly/patternfly/components/Select/select.css';
import { css } from '@patternfly/react-styles';
import PropTypes from 'prop-types';
import { SelectContext } from './selectConstants';

const propTypes = {
  /** the value for the option */
  children: PropTypes.string,
  /** additional classes added to the Select Option */
  className: PropTypes.string,
  /** the value for the option */
  value: PropTypes.string,
  /** flag indicating if the option is disabled */
  isDisabled: PropTypes.bool,
  /** flag indicating if the option acts as a placeholder */
  isPlaceholder: PropTypes.bool,
  /** Optional on click callback */
  onClick: PropTypes.func,
  /** Callback for ref tracking */
  sendRef: PropTypes.func,
  /** Additional props are spread to the container <button> */
  '': PropTypes.any
};

const defaultProps = {
  children: null,
  className: '',
  value: null,
  isDisabled: false,
  isPlaceholder: false,
  onClick: Function.prototype
};

class SelectOption extends React.Component {
  ref = React.createRef();

  componentDidMount() {
    this.props.sendRef(this.ref, this.props.index);
  }

  onKeyDown = event => {
    if (event.key === 'Tab') return;
    event.preventDefault();
    if (event.key === 'ArrowUp') {
      this.props.keyHandler(this.props.index, 'up');
    } else if (event.key === 'ArrowDown') {
      this.props.keyHandler(this.props.index, 'down');
    } else if (event.key === 'Enter') {
      this.ref.current.click && this.ref.current.click();
    }
  };

  render() {
    const {
      children,
      className,
      value,
      onClick,
      isDisabled,
      isPlaceholder,
      selected,
      sendRef,
      keyHandler,
      index,
      ...props
    } = this.props;
    return (
      <SelectContext.Consumer>
        {onSelect => (
          <li>
            <button
              {...props}
              className={css(
                styles.selectMenuItem,
                selected && styles.selectMenuItemMatch,
                isDisabled && styles.modifiers.disabled,
                className
              )}
              onClick={event => {
                if (!isDisabled) {
                  onClick && onClick(event);
                  onSelect && onSelect(event, value || children, isPlaceholder);
                }
              }}
              role="option"
              aria-selected={selected || null}
              ref={this.ref}
              onKeyDown={this.onKeyDown}
            >
              {value || children}
            </button>
          </li>
        )}
      </SelectContext.Consumer>
    );
  }
}

SelectOption.propTypes = propTypes;
SelectOption.defaultProps = defaultProps;

export default SelectOption;
