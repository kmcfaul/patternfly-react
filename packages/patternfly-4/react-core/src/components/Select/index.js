export { default as Select } from './Select';
export { default as SelectOption } from './SelectOption';
export { default as SelectToggle } from './SelectToggle';
export { SelectContext, SelectVariant } from './selectConstants';
