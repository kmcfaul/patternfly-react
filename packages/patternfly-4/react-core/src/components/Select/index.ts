export * from './Select';
export * from './SelectOption';
export * from './selectConstants';
export * from './SelectGroup';
