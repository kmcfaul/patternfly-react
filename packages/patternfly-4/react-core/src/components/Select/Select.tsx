import * as React from 'react';
import styles from '@patternfly/react-styles/css/components/Select/select';
import badgeStyles from '@patternfly/react-styles/css/components/Badge/badge';
import formStyles from '@patternfly/react-styles/css/components/FormControl/form-control';
import buttonStyles from '@patternfly/react-styles/css/components/Button/button';
import { css } from '@patternfly/react-styles';
import { TimesCircleIcon } from '@patternfly/react-icons';
import { SelectMenu } from './SelectMenu';
import { SelectOption } from './SelectOption';
import { SelectToggle } from './SelectToggle';
import { SelectContext, SelectVariant } from './selectConstants';
import { Chip, ChipGroup } from '../ChipGroup';
import { keyHandler, getNextIndex } from '../../helpers/util';
import { Omit } from '../../helpers/typeUtils';

// seed for the aria-labelledby ID
let currentId = 0;

export interface SelectProps extends Omit<React.HTMLProps<HTMLDivElement>, 'onSelect' | 'ref' | 'checked' | 'selected'> {
  /** Content rendered inside the Select */
  children: React.ReactElement[];
  /** Classes applied to the root of the Select */
  className?: string;
  /** Flag to indicate if select is expanded */
  isExpanded?: boolean;
  /** Flag to indicate if select options are grouped */
  isGrouped?: boolean;
  /** Title text of Select */
  placeholderText?: string | React.ReactNode;
  /** Selected item */
  selections?: string[] | string;
  /** Id for select toggle element */
  toggleId?: string;
  /** Adds accessible text to Select */
  'aria-label'?: string;
  /** Id of label for the Select aria-labelledby */
  ariaLabelledBy?: string;
  /** Label for input field of type ahead select variants */
  ariaLabelTypeAhead?: string;
  /** Label for clear selection button of type ahead select variants */
  ariaLabelClear?: string;
  /** Label for toggle of type ahead select variants */
  ariaLabelToggle?: string;
  /** Label for remove chip button of multiple type ahead select variant */
  ariaLabelRemove?: string;
  /** Callback for selection behavior */
  onSelect?: (event: React.MouseEvent | React.ChangeEvent, value: string, isPlaceholder?: boolean) => void;
  /** Callback for toggle button behavior */
  onToggle: (isExpanded: boolean) => void;
  /** Callback for typeahead clear button */
  onClear?: (event: React.MouseEvent) => void;
  /** Variant of rendered Select */
  variant?: 'single' | 'checkbox' | 'typeahead' | 'typeaheadmulti';
  /** Width of the select container as a number of px or string percentage */
  width?: string | number;
}

export interface SelectState {
  openedOnEnter: boolean;
  typeaheadInputValue: string;
  typeaheadActiveChild?: HTMLElement;
  typeaheadFilteredChildren: React.ReactNode[];
  typeaheadCurrIndex: number;
}

export class Select extends React.Component<SelectProps, SelectState> {
  private parentRef = React.createRef<HTMLDivElement>();
  private refCollection: HTMLElement[] = [];

  static defaultProps = {
    children: [] as React.ReactElement[],
    className: '',
    toggleId: null as string,
    isExpanded: false,
    isGrouped: false,
    'aria-label': '',
    ariaLabelledBy: '',
    ariaLabelTypeAhead: '',
    ariaLabelClear: 'Clear all',
    ariaLabelToggle: 'Options menu',
    ariaLabelRemove: 'Remove',
    selections: '',
    placeholderText: '',
    variant: SelectVariant.single,
    width: '',
    onClear: Function.prototype
  };

  state = {
    openedOnEnter: false,
    typeaheadInputValue: '',
    typeaheadActiveChild: null as HTMLElement,
    typeaheadFilteredChildren: React.Children.toArray(this.props.children),
    typeaheadCurrIndex: -1
  };

  componentDidUpdate = (prevProps: SelectProps, prevState: SelectState) => {
    if (!prevState.openedOnEnter && this.state.openedOnEnter) {
      this.refCollection[0].focus();
    }
  };

  onEnter = () => {
    this.setState({ openedOnEnter: true });
  };

  onClose = () => {
    this.setState({
      openedOnEnter: false,
      typeaheadInputValue: null,
      typeaheadActiveChild: null,
      typeaheadFilteredChildren: React.Children.toArray(this.props.children),
      typeaheadCurrIndex: -1
    });
  };

  onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    let input: RegExp;
    try {
      input = new RegExp(e.target.value, 'i');
    } catch (err) {
      input = new RegExp(e.target.value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'i');
    }
    const typeaheadFilteredChildren =
      e.target.value !== ''
        ? React.Children.toArray(this.props.children).filter(
            (child: React.ReactNode) => (child as React.ReactElement).props.value.search(input) === 0
          )
        : React.Children.toArray(this.props.children);
    if (typeaheadFilteredChildren.length === 0) {
      typeaheadFilteredChildren.push(<SelectOption isDisabled key={0} value="No results found" />);
    }
    this.setState({
      typeaheadInputValue: e.target.value,
      typeaheadCurrIndex: -1,
      typeaheadFilteredChildren,
      typeaheadActiveChild: null
    });
    this.refCollection = [];
  };

  onClick = (e: React.MouseEvent) => {
    e.stopPropagation();
  };

  clearSelection = (e: React.MouseEvent) => {
    e.stopPropagation();
    this.setState({
      typeaheadInputValue: '',
      typeaheadActiveChild: null,
      typeaheadFilteredChildren: React.Children.toArray(this.props.children),
      typeaheadCurrIndex: -1
    });
  };

  extendTypeaheadChildren(typeaheadActiveChild: HTMLElement) {
    return this.state.typeaheadFilteredChildren.map((child: React.ReactNode) =>
      React.cloneElement(child as React.ReactElement, {
        isFocused: typeaheadActiveChild && typeaheadActiveChild.innerText === (child as React.ReactElement).props.value
      })
    );
  }

  sendRef = (ref: React.ReactNode, index: number) => {
    this.refCollection[index] = ref as HTMLElement;
  };

  handleArrowKeys = (index: number, position: string) => {
    keyHandler(index, position, this.refCollection, this.refCollection);
  };

  handleTypeaheadKeys = (position: string) => {
    const { isExpanded, onSelect } = this.props;
    const { typeaheadActiveChild, typeaheadCurrIndex } = this.state;
    if (isExpanded) {
      if (position === 'enter' && (typeaheadActiveChild || this.refCollection[0])) {
        this.setState({
          typeaheadInputValue:
            (typeaheadActiveChild && typeaheadActiveChild.innerText) || this.refCollection[0].innerText
        });
        onSelect(null, (typeaheadActiveChild && typeaheadActiveChild.innerText) || this.refCollection[0].innerText);
      } else {
        let nextIndex;
        if (typeaheadCurrIndex === -1 && position === 'down') {
          nextIndex = 0;
        } else if (typeaheadCurrIndex === -1 && position === 'up') {
          nextIndex = this.refCollection.length - 1;
        } else {
          nextIndex = getNextIndex(typeaheadCurrIndex, position, this.refCollection);
        }
        this.setState({
          typeaheadCurrIndex: nextIndex,
          typeaheadActiveChild: this.refCollection[nextIndex],
          typeaheadInputValue: this.refCollection[nextIndex].innerText
        });
      }
    }
  };

  render() {
    const {
      children,
      className,
      variant,
      onToggle,
      onSelect,
      onClear,
      toggleId,
      isExpanded,
      isGrouped,
      selections,
      ariaLabelledBy,
      ariaLabelTypeAhead,
      ariaLabelClear,
      ariaLabelToggle,
      ariaLabelRemove,
      'aria-label': ariaLabel,
      placeholderText,
      width,
      ...props
    } = this.props;
    const { openedOnEnter, typeaheadInputValue, typeaheadActiveChild } = this.state;
    const selectToggleId = toggleId || `pf-toggle-id-${currentId++}`;
    let childPlaceholderText = null;
    if (!selections && !placeholderText) {
      const childPlaceholder = React.Children.toArray(children.filter(child => child.props.isPlaceholder === true));
      childPlaceholderText =
        (childPlaceholder[0] && childPlaceholder[0].props.value) || (children[0] && children[0].props.value);
    }
    let selectedChips = null;
    if (variant === SelectVariant.typeaheadMulti) {
      selectedChips = (
        <ChipGroup>
          {selections &&
            (selections as string[]).map(item => (
              <Chip key={item} onClick={e => onSelect(e, item)} closeBtnAriaLabel={ariaLabelRemove}>
                {item}
              </Chip>
            ))}
        </ChipGroup>
      );
    }

    return (
      <div
        className={css(styles.select, isExpanded && styles.modifiers.expanded, className)}
        ref={this.parentRef}
        style={{ width }}
      >
        <SelectContext.Provider value={{ onSelect, onClose: this.onClose, variant }}>
          <SelectToggle
            id={selectToggleId}
            parentRef={this.parentRef}
            isExpanded={isExpanded}
            onToggle={onToggle}
            onEnter={this.onEnter}
            onClose={this.onClose}
            ariaLabelledBy={`${ariaLabelledBy || ''} ${selectToggleId}`}
            variant={variant}
            ariaLabelToggle={ariaLabelToggle}
            handleTypeaheadKeys={this.handleTypeaheadKeys}
          >
            {variant === SelectVariant.single && (
              <div className={css(styles.selectToggleWrapper)}>
                <span className={css(styles.selectToggleText)}>
                  {selections || placeholderText || childPlaceholderText}
                </span>
              </div>
            )}
            {variant === SelectVariant.checkbox && (
              <React.Fragment>
                <div className={css(styles.selectToggleWrapper)}>
                  <span className={css(styles.selectToggleText)}>{placeholderText}</span>
                  {selections && selections.length > 0 && (
                    <div className={css(styles.selectToggleBadge)}>
                      <span className={css(badgeStyles.badge, badgeStyles.modifiers.read)}>{selections.length}</span>
                    </div>
                  )}
                </div>
              </React.Fragment>
            )}
            {variant === SelectVariant.typeahead && (
              <React.Fragment>
                <div className={css(styles.selectToggleWrapper)}>
                  <input
                    className={css(formStyles.formControl, styles.selectToggleTypeahead)}
                    aria-activedescendant={typeaheadActiveChild && typeaheadActiveChild.id}
                    id="select-typeahead"
                    aria-label={ariaLabelTypeAhead}
                    placeholder={placeholderText as string}
                    value={typeaheadInputValue !== null ? typeaheadInputValue : selections || ''}
                    type="text"
                    onChange={this.onChange}
                    autoComplete="off"
                  />
                </div>
                {selections && (
                  <button
                    className={css(buttonStyles.button, buttonStyles.modifiers.plain, styles.selectToggleClear)}
                    onClick={e => {
                      this.clearSelection(e);
                      onClear(e);
                    }}
                    aria-label={ariaLabelClear}
                  >
                    <TimesCircleIcon aria-hidden />
                  </button>
                )}
              </React.Fragment>
            )}
            {variant === SelectVariant.typeaheadMulti && (
              <React.Fragment>
                <div className={css(styles.selectToggleWrapper)}>
                  {selections && selections.length > 0 && selectedChips}
                  <input
                    className={css(formStyles.formControl, styles.selectToggleTypeahead)}
                    aria-activedescendant={typeaheadActiveChild && typeaheadActiveChild.id}
                    id="select-multi-typeahead-typeahead"
                    aria-label={ariaLabelTypeAhead}
                    placeholder={placeholderText as string}
                    value={typeaheadInputValue !== null ? typeaheadInputValue : ''}
                    type="text"
                    onChange={this.onChange}
                    autoComplete="off"
                  />
                </div>
                {selections && selections.length > 0 && (
                  <button
                    className={css(buttonStyles.button, buttonStyles.modifiers.plain, styles.selectToggleClear)}
                    onClick={e => {
                      this.clearSelection(e);
                      onClear(e);
                    }}
                    aria-label={ariaLabelClear}
                  >
                    <TimesCircleIcon aria-hidden />
                  </button>
                )}
              </React.Fragment>
            )}
          </SelectToggle>
          {variant === SelectVariant.single && isExpanded && (
            <SelectMenu
              {...props}
              selected={selections}
              openedOnEnter={openedOnEnter}
              aria-label={ariaLabel}
              aria-labelledby={ariaLabelledBy}
              sendRef={this.sendRef}
              keyHandler={this.handleArrowKeys}
            >
              {children}
            </SelectMenu>
          )}
          {variant === SelectVariant.checkbox && isExpanded && (
            <SelectMenu
              {...props}
              checked={selections as string[]}
              aria-label={ariaLabel}
              aria-labelledby={ariaLabelledBy}
              isGrouped={isGrouped}
              sendRef={this.sendRef}
              keyHandler={this.handleArrowKeys}
            >
              {children}
            </SelectMenu>
          )}
          {(variant === SelectVariant.typeahead || variant === SelectVariant.typeaheadMulti) && isExpanded && (
            <SelectMenu
              {...props}
              selected={selections}
              openedOnEnter={openedOnEnter}
              aria-label={ariaLabel}
              aria-labelledby={ariaLabelledBy}
              sendRef={this.sendRef}
              keyHandler={this.handleArrowKeys}
            >
              {this.extendTypeaheadChildren(typeaheadActiveChild)}
            </SelectMenu>
          )}
        </SelectContext.Provider>
      </div>
    );
  }
}
