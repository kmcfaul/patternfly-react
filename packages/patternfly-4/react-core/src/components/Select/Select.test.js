import React from 'react';
import { mount } from 'enzyme';
import Select from './Select';
import SelectOption from './SelectOption';

const selectOptions = [
  <SelectOption value="Mr" />,
  <SelectOption value="Mrs" />,
  <SelectOption value="Ms" />,
  <SelectOption value="Other" />
];

describe('select', () => {
  describe('single select', () => {
    test('renders closed successfully', () => {
      const view = mount(<Select variant="single">{selectOptions}</Select>);
      expect(view).toMatchSnapshot();
    });

    test('renders expanded successfully', () => {
      const view = mount(
        <Select variant="single" isExpanded>
          {selectOptions}
        </Select>
      );
      expect(view).toMatchSnapshot();
    });

    test('renders with selectOptions parameter', () => {
      const view = mount(<Select variant="single" selectOptions={selectOptions} />);
      expect(view).toMatchSnapshot();
    });
  });
});

describe('API', () => {
  test('click on item', () => {
    const mockToggle = jest.fn();
    const mockSelect = jest.fn();
    const view = mount(
      <Select variant="single" onToggle={mockToggle} onSelect={mockSelect} selectOptions={selectOptions} isExpanded />
    );
    view
      .find('button')
      .first()
      .simulate('click');
    expect(mockToggle.mock.calls).toHaveLength(0);
    expect(mockSelect.mock.calls).toHaveLength(1);
  });

  test('selectOptions and children console error ', () => {
    const myMock = jest.fn();
    global.console = { error: myMock };
    mount(
      <Select variant="single" selectOptions={selectOptions} isExpanded>
        <div> child test </div>
        <div> child test </div>
      </Select>
    );
    expect(myMock).toBeCalled();
  });

  test('selectOptions only, no console error ', () => {
    const myMock = jest.fn();
    global.console = { error: myMock };
    mount(<Select variant="single" selectOptions={selectOptions} isExpanded />);
    expect(myMock).not.toBeCalled();
  });

  test('children only, no console error', () => {
    const myMock = jest.fn();
    global.console = { error: myMock };
    mount(
      <Select variant="single" isExpanded>
        {selectOptions}
      </Select>
    );
    expect(myMock).not.toBeCalled();
  });
});
