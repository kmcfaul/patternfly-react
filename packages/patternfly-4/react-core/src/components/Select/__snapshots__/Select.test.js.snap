// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select single select renders closed successfully 1`] = `
.pf-c-select__toggle-text {
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}
.pf-c-select__toggle-wrapper {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: flex-start;
  max-width: calc(100% - 1.5rem);
  white-space: normal;
}
.pf-c-select__toggle-arrow {
  display: block;
}
.pf-c-select__toggle-arrow {
  display: block;
}
.pf-c-select__toggle {
  display: flex;
  position: relative;
  align-items: center;
  justify-content: space-between;
  min-width: 44px;
  max-width: 100%;
  padding: 0.375rem 0.5rem 0.375rem 0.5rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #282d33;
  white-space: nowrap;
  cursor: pointer;
  background-color: #ffffff;
}
.pf-c-select {
  display: inline-block;
  position: relative;
  max-width: 100%;
}

<Select
  className=""
  isExpanded={false}
  placeholderText="Please choose one"
  selectOptions={null}
  selections={null}
  variant="single"
>
  <div
    className="pf-c-select"
  >
    <SelectToggle
      className=""
      iconComponent={[Function]}
      id="pf-toggle-id-0"
      isActive={false}
      isExpanded={false}
      isFocused={false}
      isHovered={false}
      isPlain={false}
      onClose={[Function]}
      onEnter={[Function]}
      onToggle={[Function]}
      parentRef={null}
    >
      <div
        aria-expanded={false}
        aria-haspopup="listbox"
        className="pf-c-select__toggle"
        id="pf-toggle-id-0"
        onClick={[Function]}
        onClose={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
        type="button"
      >
        <div
          className="pf-c-select__toggle-wrapper"
        >
          <div
            className="pf-c-select__toggle-text"
          >
            Please choose one
          </div>
        </div>
        <CaretDownIcon
          className="pf-c-select__toggle-arrow"
          color="currentColor"
          size="sm"
          title={null}
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            className="pf-c-select__toggle-arrow"
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              transform=""
            />
          </svg>
        </CaretDownIcon>
      </div>
    </SelectToggle>
  </div>
</Select>
`;

exports[`select single select renders expanded successfully 1`] = `
.pf-c-select__toggle-text {
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}
.pf-c-select__toggle-wrapper {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: flex-start;
  max-width: calc(100% - 1.5rem);
  white-space: normal;
}
.pf-c-select__toggle-arrow {
  display: block;
}
.pf-c-select__toggle-arrow {
  display: block;
}
.pf-c-select__toggle {
  display: flex;
  position: relative;
  align-items: center;
  justify-content: space-between;
  min-width: 44px;
  max-width: 100%;
  padding: 0.375rem 0.5rem 0.375rem 0.5rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #282d33;
  white-space: nowrap;
  cursor: pointer;
  background-color: #ffffff;
}
.pf-c-select__menu-item {
  display: block;
  width: 100%;
  padding: 0.5rem 1rem 0.5rem 1rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #282d33;
  text-align: left;
  white-space: nowrap;
  background-color: #ededed;
  text-decoration: none;
}
.pf-c-select__menu-item {
  display: block;
  width: 100%;
  padding: 0.5rem 1rem 0.5rem 1rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #282d33;
  text-align: left;
  white-space: nowrap;
  background-color: #ededed;
  text-decoration: none;
}
.pf-c-select__menu-item {
  display: block;
  width: 100%;
  padding: 0.5rem 1rem 0.5rem 1rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #282d33;
  text-align: left;
  white-space: nowrap;
  background-color: #ededed;
  text-decoration: none;
}
.pf-c-select__menu-item {
  display: block;
  width: 100%;
  padding: 0.5rem 1rem 0.5rem 1rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #282d33;
  text-align: left;
  white-space: nowrap;
  background-color: #ededed;
  text-decoration: none;
}
.pf-c-select__menu {
  display: block;
  position: absolute;
  z-index: 100;
  min-width: 100%;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  background-color: #ffffff;
  background-clip: padding-box;
  border: 1px solid transparent;
  box-shadow: 0 0.0625rem 0.0625rem 0rem rgba(3, 3, 3, 0.05), 0 0.25rem 0.5rem 0.25rem rgba(3, 3, 3, 0.06);
}
.pf-c-select.pf-m-expanded {
  display: inline-block;
  position: relative;
  max-width: 100%;
}

<Select
  className=""
  isExpanded={true}
  placeholderText="Please choose one"
  selectOptions={null}
  selections={null}
  variant="single"
>
  <div
    className="pf-c-select pf-m-expanded"
  >
    <SelectToggle
      className=""
      iconComponent={[Function]}
      id="pf-toggle-id-1"
      isActive={false}
      isExpanded={true}
      isFocused={false}
      isHovered={false}
      isPlain={false}
      onClose={[Function]}
      onEnter={[Function]}
      onToggle={[Function]}
      parentRef={null}
    >
      <div
        aria-expanded={true}
        aria-haspopup="listbox"
        className="pf-c-select__toggle"
        id="pf-toggle-id-1"
        onClick={[Function]}
        onClose={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
        type="button"
      >
        <div
          className="pf-c-select__toggle-wrapper"
        >
          <div
            className="pf-c-select__toggle-text"
          >
            Please choose one
          </div>
        </div>
        <CaretDownIcon
          className="pf-c-select__toggle-arrow"
          color="currentColor"
          size="sm"
          title={null}
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            className="pf-c-select__toggle-arrow"
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              transform=""
            />
          </svg>
        </CaretDownIcon>
      </div>
    </SelectToggle>
    <SingleSelect
      className=""
      isExpanded={false}
      openedOnEnter={false}
      selected={null}
    >
      <ul
        className="pf-c-select__menu"
        role="listbox"
      >
        <SelectOption
          className=""
          index={0}
          isDisabled={false}
          key=".0"
          keyHandler={[Function]}
          onClick={[Function]}
          selected={false}
          sendRef={[Function]}
          value="Mr"
        >
          <li>
            <button
              aria-selected={null}
              className="pf-c-select__menu-item"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
            >
              Mr
            </button>
          </li>
        </SelectOption>
        <SelectOption
          className=""
          index={1}
          isDisabled={false}
          key=".1"
          keyHandler={[Function]}
          onClick={[Function]}
          selected={false}
          sendRef={[Function]}
          value="Mrs"
        >
          <li>
            <button
              aria-selected={null}
              className="pf-c-select__menu-item"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
            >
              Mrs
            </button>
          </li>
        </SelectOption>
        <SelectOption
          className=""
          index={2}
          isDisabled={false}
          key=".2"
          keyHandler={[Function]}
          onClick={[Function]}
          selected={false}
          sendRef={[Function]}
          value="Ms"
        >
          <li>
            <button
              aria-selected={null}
              className="pf-c-select__menu-item"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
            >
              Ms
            </button>
          </li>
        </SelectOption>
        <SelectOption
          className=""
          index={3}
          isDisabled={false}
          key=".3"
          keyHandler={[Function]}
          onClick={[Function]}
          selected={false}
          sendRef={[Function]}
          value="Other"
        >
          <li>
            <button
              aria-selected={null}
              className="pf-c-select__menu-item"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
            >
              Other
            </button>
          </li>
        </SelectOption>
      </ul>
    </SingleSelect>
  </div>
</Select>
`;

exports[`select single select renders with selectOptions parameter 1`] = `
.pf-c-select__toggle-text {
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}
.pf-c-select__toggle-wrapper {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: flex-start;
  max-width: calc(100% - 1.5rem);
  white-space: normal;
}
.pf-c-select__toggle-arrow {
  display: block;
}
.pf-c-select__toggle-arrow {
  display: block;
}
.pf-c-select__toggle {
  display: flex;
  position: relative;
  align-items: center;
  justify-content: space-between;
  min-width: 44px;
  max-width: 100%;
  padding: 0.375rem 0.5rem 0.375rem 0.5rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #282d33;
  white-space: nowrap;
  cursor: pointer;
  background-color: #ffffff;
}
.pf-c-select {
  display: inline-block;
  position: relative;
  max-width: 100%;
}

<Select
  className=""
  isExpanded={false}
  placeholderText="Please choose one"
  selectOptions={
    Array [
      <SelectOption
        className=""
        isDisabled={false}
        onClick={[Function]}
        value="Mr"
      />,
      <SelectOption
        className=""
        isDisabled={false}
        onClick={[Function]}
        value="Mrs"
      />,
      <SelectOption
        className=""
        isDisabled={false}
        onClick={[Function]}
        value="Ms"
      />,
      <SelectOption
        className=""
        isDisabled={false}
        onClick={[Function]}
        value="Other"
      />,
    ]
  }
  selections={null}
  variant="single"
>
  <div
    className="pf-c-select"
  >
    <SelectToggle
      className=""
      iconComponent={[Function]}
      id="pf-toggle-id-2"
      isActive={false}
      isExpanded={false}
      isFocused={false}
      isHovered={false}
      isPlain={false}
      onClose={[Function]}
      onEnter={[Function]}
      onToggle={[Function]}
      parentRef={null}
    >
      <div
        aria-expanded={false}
        aria-haspopup="listbox"
        className="pf-c-select__toggle"
        id="pf-toggle-id-2"
        onClick={[Function]}
        onClose={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
        type="button"
      >
        <div
          className="pf-c-select__toggle-wrapper"
        >
          <div
            className="pf-c-select__toggle-text"
          >
            Please choose one
          </div>
        </div>
        <CaretDownIcon
          className="pf-c-select__toggle-arrow"
          color="currentColor"
          size="sm"
          title={null}
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            className="pf-c-select__toggle-arrow"
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              transform=""
            />
          </svg>
        </CaretDownIcon>
      </div>
    </SelectToggle>
  </div>
</Select>
`;
